pkg/controllers/managedzone/suite_test.go

```diff
@@ -0,0 +1,443 @@
+			err := k8sClient.List(ctx, mzList, client.InNamespace("default"))
+			Expect(err).NotTo(HaveOccurred())
+			for _, mz := range mzList.Items {
+				err = k8sClient.Delete(ctx, &mz)
+				Expect(err).NotTo(HaveOccurred())
+			}
+		})
+
+		It("should accept a managed zone for this controller and allow deletion", func() {
+			Expect(k8sClient.Create(ctx, managedZone)).To(BeNil())
+
+			createdMZ := &v1alpha1.ManagedZone{}
+
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, client.ObjectKey{Namespace: managedZone.Namespace, Name: managedZone.Name}, createdMZ)
+				return err == nil
+			}, EventuallyTimeoutMedium, TestRetryIntervalMedium).Should(BeTrue())
+
+			Expect(k8sClient.Delete(ctx, managedZone)).To(BeNil())
+
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, client.ObjectKey{Namespace: managedZone.Namespace, Name: managedZone.Name}, createdMZ)
+				return errors.IsNotFound(err)
+			}, EventuallyTimeoutMedium, TestRetryIntervalMedium).Should(BeTrue())
+		})
+
+		It("should accept a managed zone and permit related hosts and deny unrelated hosts", func() {
+			Expect(k8sClient.Create(ctx, managedZone)).To(BeNil())
+
+			createdMZ := &v1alpha1.ManagedZone{}
+
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, client.ObjectKey{Namespace: managedZone.Namespace, Name: managedZone.Name}, createdMZ)
+				return err == nil
+			}, EventuallyTimeoutMedium, TestRetryIntervalMedium).Should(BeTrue())
+
+			Expect(k8sClient.Create(ctx, gatewayClass)).To(BeNil())
+			createdGWClass := &gatewayv1beta1.GatewayClass{}
+
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, client.ObjectKey{Name: gatewayClass.Name}, createdGWClass)
+				return err == nil
+			}, EventuallyTimeoutMedium, TestRetryIntervalMedium).Should(BeTrue())
+
+			Expect(k8sClient.Create(ctx, gateway)).To(BeNil())
+			createdGW := &gatewayv1beta1.Gateway{}
+
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, client.ObjectKey{Name: gateway.Name, Namespace: gateway.Namespace}, createdGW)
+				return err == nil
+			}, EventuallyTimeoutMedium, TestRetryIntervalMedium).Should(BeTrue())
+
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, client.ObjectKey{Name: gateway.Name, Namespace: gateway.Namespace}, createdGW)
+				if err != nil {
+					return false
+				}
+				err = k8sClient.Get(ctx, client.ObjectKey{Namespace: managedZone.Namespace, Name: managedZone.Name}, createdMZ)
+				if err != nil {
+					return false
+				}
+
+				mzIsReady := false
+				for _, condition := range createdMZ.Status.Conditions {
+					if condition.Type == common.ReadyStatusConditionType && condition.Status == metav1.ConditionTrue {
+						mzIsReady = true
+					}
+				}
+
+				gwIsProgrammed := false
+				for _, condition := range createdGW.Status.Conditions {
+					if condition.Type == "Programmed" && condition.Status == metav1.ConditionTrue {
+						gwIsProgrammed = true
+					}
+				}
+
+				return mzIsReady && gwIsProgrammed
```

Comment:
I'm not sure why, but it seems odd to me that the existence of a manage zone has any influence on whether a gateway is programmed or not. Is this only true if we have at least one host that has a corresponding managed zone that we can create a record in?

Reply:
yes, exactly.